sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,140 @@
+  mmio_region_write32(config->base_addr, HMAC_CFG(0), device_config);
+
+  // Clear INTER_STATE.
+  mmio_region_write32(config->base_addr, HMAC_INTR_STATE(0),
+                      1 << HMAC_INTR_ENABLE_HMAC_DONE |
+                          1 << HMAC_INTR_STATE_FIFO_FULL |
+                          1 << HMAC_INTR_STATE_HMAC_ERR);
+
+  // Set the byte-order of the input message.
+  if (config->endian_swap) {
+    device_config |= 1 << HMAC_CFG_ENDIAN_SWAP;
+  }
+  // Set the byte-order of the digest.
+  if (config->digest_swap) {
+    device_config |= 1 << HMAC_CFG_DIGEST_SWAP;
+  }
+
+  // Set HMAC engine to HMAC or SHA256-only mode.
+  if (config->mode == DIF_HMAC_OP_HMAC) {
+    device_config |= 1 << HMAC_CFG_HMAC_EN;
+
+    // Set the key making no assumptions about register locations.
+    mmio_region_write32(config->base_addr, HMAC_KEY0(0), config->key[0]);
+    mmio_region_write32(config->base_addr, HMAC_KEY1(0), config->key[1]);
+    mmio_region_write32(config->base_addr, HMAC_KEY2(0), config->key[2]);
+    mmio_region_write32(config->base_addr, HMAC_KEY3(0), config->key[3]);
+    mmio_region_write32(config->base_addr, HMAC_KEY4(0), config->key[4]);
+    mmio_region_write32(config->base_addr, HMAC_KEY5(0), config->key[5]);
+    mmio_region_write32(config->base_addr, HMAC_KEY6(0), config->key[6]);
+    mmio_region_write32(config->base_addr, HMAC_KEY7(0), config->key[7]);
+  }
+  device_config |= 1 << HMAC_CFG_SHA_EN;
+
+  // Write the configuration.
+  mmio_region_write32(config->base_addr, HMAC_CFG(0), device_config);
+
+  // Begin SHA or HMAC operation.
+  mmio_region_nonatomic_set_bit32(config->base_addr, HMAC_CMD(0),
+                                  HMAC_CMD_HASH_START);
+
+  // Setup state tracking.
+  hmac_out->base_addr = config->base_addr;
+}
+
+size_t dif_hmac_push_fifo(const dif_hmac_t *hmac, void *data, size_t len) {
+  const uint32_t *wp = (uint32_t *)data;
+  uint32_t words_remaining = len / sizeof(uint32_t);
+  uint8_t bytes_remaining = len % sizeof(uint32_t);
+
+  // Write data until we either finish or the FIFO fills up.
+  while ((words_remaining || bytes_remaining) && fifo_avail(hmac) > 0) {
+    if (words_remaining > 0) {
+      // Write full words to the FIFO
+      mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO(0), *wp++);
+      words_remaining--;
+    } else {
+      // Write the remaining bytes to the FIFO.
+      uint8_t *bp = (uint8_t *)wp;
+
+      for (; bytes_remaining > 0; bytes_remaining--) {
+        mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO(0), *bp++);
+      }
+    }
+  }
+
+  return len - words_remaining * sizeof(uint32_t) - bytes_remaining;
+}
+
+uint64_t dif_hmac_get_msg_len(const dif_hmac_t *hmac) {
+  uint64_t msg_lower =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_LOWER(0));
+  uint64_t msg_upper =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_UPPER(0));
+
+  return (msg_upper << (sizeof(uint32_t) * 8)) | msg_lower;
```

Comment:
I wonder if we want to add this type of operation to a header in libbase (e.g., u64_from_u32s())...